daemon/listeners/group_unix.go

```diff
@@ -4,27 +4,16 @@ package listeners // import "github.com/docker/docker/daemon/listeners"
 
 import (
 	"fmt"
+	"github.com/docker/docker/pkg/idtools"
 	"strconv"
-
-	"github.com/opencontainers/runc/libcontainer/user"
-	"github.com/pkg/errors"
 )
 
 const defaultSocketGroup = "docker"
 
 func lookupGID(name string) (int, error) {
-	groupFile, err := user.GetGroupPath()
-	if err != nil {
-		return -1, errors.Wrap(err, "error looking up groups")
-	}
-	groups, err := user.ParseGroupFileFilter(groupFile, func(g user.Group) bool {
-		return g.Name == name || strconv.Itoa(g.Gid) == name
-	})
-	if err != nil {
-		return -1, errors.Wrapf(err, "error parsing groups for %s", name)
-	}
-	if len(groups) > 0 {
-		return groups[0].Gid, nil
+	group, err := idtools.LookupGroup(name)
+	if err == nil {
```

Comment:
Wondering if we should log the error as a warning if we both fail to find a group, and converting the `name` to an integer also fails.

(It's a bit unfortunate that "group not found", and "an error occurred looking up the group" both return an error here; https://github.com/moby/moby/blob/b3e9f7b13b0f0c414fa6253e1f17a86b2cff68b5/pkg/idtools/idtools_unix.go#L183 🤔 ideally we should have a "IsNotFound"-like error, or perhaps no error at all for that)

Reply:
Regarding the specific case of `daemon/listeners/group_unix.go`, errors get intercepted and are logged if appropriate in `listeners_linux.go` :
https://github.com/moby/moby/blob/980c2992c679ca15be0d865b8b7040dfdc28278a/daemon/listeners/listeners_linux.go#L40 . This is the only call site to `group_unix.lookupGID()`.

As for the more general matter of error handling in `idtools_unix.go`, I agree that some minor improvement could be possible, but I think the behaviour of returning an error status (whatever its exact semantic) is the proper way to go. After all, the fact that a named group doesn't exist is not necessarily an error in all use case, but callers definitely need a way to help the user identify why a group was not found, notably when using external sources... I also think that it would be desirable to move to `idtools_unix.go` the handling of "the group name is actually a numerical id" cases, as it appears (well, at least to me) to always be a desirable fallback...

Now, come to that, I have noticed that there are quite a few places in Docker's source code where `passwd` and `group` files are crawled using various ad-hoc strategies. If more efforts is to be put in `idtools_unix.go`, then maybe it would be wise to consider using those functions everywhere instead of having custom code there and there...

By the way, I tried to retain the original behaviour as much as possible, in the hope that it would make the fix more acceptable for inclusion, essentially because it is my first contribution here. Still, I can certainly put a little more work in it if you are willing to give me some hints on what is to be done.

Reply:
Ok, after some more checks, it turns out that there are almost no case left of ad-hoc parsing. I only found [`oci_linux.go`](https://github.com/moby/moby/blob/deac65c929504acd7de6567d0422ead7e87441ac/daemon/oci_linux.go#L165-L208) and [`internals_linux.go`](https://github.com/moby/moby/blob/b3e9f7b13b0f0c414fa6253e1f17a86b2cff68b5/builder/dockerfile/internals_linux.go#L52-L88), and I don't think that supporting external user sources for the second would be desirable.